chacha20_poly1305/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+}
+
+/// Encrypt and decrypt content along with a authentication tag.
+pub struct ChaCha20Poly1305 {
+    key: Key,
+    nonce: Nonce,
+}
+
+impl ChaCha20Poly1305 {
+    /// Make a new instance of a ChaCha20Poly1305 AEAD.
+    pub const fn new(key: Key, nonce: Nonce) -> Self {
+        ChaCha20Poly1305 { key, nonce }
+    }
+
+    /// Encrypt content in place and return the Poly1305 16-byte authentication tag.
+    ///
+    /// # Arguments
+    ///
+    /// - `content` - Plaintext to be encrypted in place.
+    /// - `aad`     - Optional metadata covered by the authentication tag.
+    ///
+    /// # Returns
+    ///
+    /// The 16-byte authentication tag.
+    pub fn encrypt(self, content: &mut [u8], aad: Option<&[u8]>) -> [u8; 16] {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 1);
+        chacha.apply_keystream(content);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        poly.input(aad);
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.input(&ZEROES[0..(16 - aad_overflow)]);
+        }
+
+        poly.input(content);
+        let text_overflow = content.len() % 16;
+        if text_overflow > 0 {
+            poly.input(&ZEROES[0..(16 - text_overflow)]);
+        }
+
+        let len_buffer = encode_lengths(aad.len() as u64, content.len() as u64);
+        poly.input(&len_buffer);
+        poly.tag()
+    }
+
+    /// Decrypt the ciphertext in place if authentication tag is correct.
+    ///
+    /// # Arguments
+    ///
+    /// - `content` - Ciphertext to be decrypted in place.
+    /// - `tag`     - 16-byte authentication tag.
+    /// - `aad`     - Optional metadata covered by the authentication tag.
+    pub fn decrypt(
+        self,
+        content: &mut [u8],
+        tag: [u8; 16],
+        aad: Option<&[u8]>,
+    ) -> Result<(), Error> {
+        let mut chacha = ChaCha20::new_from_block(self.key, self.nonce, 0);
+        let keystream = chacha.get_keystream(0);
+        let mut poly = Poly1305::new(keystream[..32].try_into().expect("infallible conversion"));
+        let aad = aad.unwrap_or(&[]);
+        poly.input(aad);
+        // AAD and ciphertext are padded if not 16-byte aligned.
+        let aad_overflow = aad.len() % 16;
+        if aad_overflow > 0 {
+            poly.input(&ZEROES[0..(16 - aad_overflow)]);
+        }
+        poly.input(content);
+        let msg_overflow = content.len() % 16;
+        if msg_overflow > 0 {
+            poly.input(&ZEROES[0..(16 - msg_overflow)]);
+        }
+
+        let len_buffer = encode_lengths(aad.len() as u64, content.len() as u64);
+        poly.input(&len_buffer);
+        let derived_tag = poly.tag();
+        if derived_tag.eq(&tag) {
```

Comment:
Why not use `==`?

Reply:
No reason not too, allows the ref to be implicit too. I'll switch it.